Changes to enable importing of previously exported CVS files

robertseliger opened the pull request.

mpkasp commented:
High level: looks great! Thanks for your hard work on this.

The main thing that bothers me a bit is the amount of "repeated" code in csv_headers.py - I say "repeated" in quotes because it's not verbatim repeated, but we are repeating ourselves whenever we call all of the model fields we already defined. 

Specifically, we're repeating ourselves in the extended classes of CSVHeaders, then repeating them again in their respective __init__ functions. 

Here's something I came up with that could help simplify this by removing the "common" case of a field _not_ having synonyms. WARNING! This is untested "pseudo code", but I think it could get us pretty close.

```python
class CSVHeader:
	def __init__(self, header, header_synonyms=None):
		self.header = header
		self.header_synonyms = header_synonyms

	# Creates a CSVHeader object from a Django model, appending the model name to the field name to make the header (lowercase), and empty list of synonyms
	# For example calling part_headers = CSVHeader.headers_from_model(Part, exclude_fields=['id']) would return:
	#   [CSVHeader(header='part_number_class'), CSVHeader(header='part_description'), ...]
	@classmethod
	def headers_from_model(cls, model, exclude_fields=None):
		header = []
		class_name = model.__class__.__name__.lower()
		for field in model._meta.get_fields():
			if exclude_fields is not None and field not in exclude_fields:
				headers.append(cls(header=class_name + '_' + field.name))
		return headers

	# Below to functions are to make sure this clas works happily with Python sets
	def __eq__(self, other): 
        if not isinstance(other, MyClass):
            return NotImplemented  # don't attempt to compare against unrelated types
        return self.header == other.header

    def __hash__(self):
        return hash(self.header)


# here we can change all_header_defns to a set() object to make sure our header fields are unique, a list would be fine too...
class CSVHeaders:
	all_header_defns = set()
	header_synonyms = []

	def __init__(self, csv_headers):
		for csv_header in csv_headers:
			if csv_header in self.all_header_defns
				raise CSVHeaderError(f'CSVHeaders object already contains header {csv_header.header}')
			all_header_defns.add(csv_header)

	...


# An example of implementing CSVHeaders - define the header_synonyms as a class attribute which auto-generates all of the headers 
# based on the Django model attributes, then when we initialize an instance of PartsListCSVHeaders, it generates all of the header definitions
class PartsListCSVHeaders(CSVHeaders):
	header_synonyms = [
		CSVHeader(header='part_description', header_synonyms=['desc', 'desc.']),
		CSVHeader(header='manufacturer_name', header_synonyms=['mfg_name', 'manufacturer_name', 'part_manufacturer', 'mfg', 'manufacturer', 'manufacturer name', ]),
		...
		# Stop at tolerance - no need to override with synonyms
	]

    def __init__(self):
    	part_headers = CSVHeader.headers_from_model(Part, exclude_fields=['id', 'description', ])
        part_revision_headers = CSVHeader.headers_from_model(PartRevision, exclude_fields=['id', ])
        manufacturer_headers = CSVHeader.headers_from_model(Manufacturer, exclude_fields=['id', 'name', ])
        headers = part_headers + part_revision_headers + manufacturer_headers
        super(CSVHeaders).__init__(csv_headers=headers)

```

A few things - 
- We could play around with what's a static class attribute vs initialized as an instance attribute, probably makes more sense to keep a lot of this as class attributes as we should know it all at runtime
- Could consider implementing new version of "get_from_dict" in the CSVHeader class
- Could consider implementing equivalent version of "get_val_from_row" in the CSVHeaders class

Thanks again for your hard work. Let me know your thoughts.

robertseliger commented:
Here are some of the use cases I ran into:

 
A CSV column header may have multiple synonymous names.


For a column that is populated from a model object field, at least one of a header’s synonyms should match the corresponding model object field name 


No all model object fields are intended to be included in a CVS file.


The order in which columns are “printed” in a CSV files needs to be controlled. This order generally has nothing to do with the order in which model fields have been declared, or with alphabetization, or anything else that can be inferred. The order of columns is for user consumption (i.e., readability).


Sometimes the value for a cell in a CSV files is obtained from a model object field, sometimes it is computed. Some of the columns populated from PartBomItem are an example of the latter.


Not all columns that are imported correspond to model object fields. An example is “synopsis” which is simply ignored when imported.
 

I am all for simplification and consolidation, and I am sure that there is a way to marry the ideas you presented below with what I started. My suggestion is that we do so interactively, via a web meeting, as this has served us well in the past.

 

If you like, we can arrange a Zoom meeting this week. Is there a time that works best for you?

 
Rob
 

 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Monday, February 17, 2020 at 1:17 AM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Author <author@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 
class CSVHeader:
        def __init__(self, header, header_synonyms=None):
               self.header = header
               self.header_synonyms = header_synonyms
 
        # Creates a CSVHeader object from a Django model, appending the model name to the field name to make the header (lowercase), and empty list of synonyms
        # For example calling part_headers = CSVHeader.headers_from_model(Part, exclude_fields=['id']) would return:
        #   [CSVHeader(header='part_number_class'), CSVHeader(header='part_description'), ...]
        @classmethod
        def headers_from_model(cls, model, exclude_fields=None):
               header = []
               class_name = model.__class__.__name__.lower()
               for field in model._meta.get_fields():
                       if exclude_fields is not None and field not in exclude_fields:
                               headers.append(cls(header=class_name + '_' + field.name))
               return headers
 
        # Below to functions are to make sure this clas works happily with Python sets
        def __eq__(self, other): 
        if not isinstance(other, MyClass):
            return NotImplemented  # don't attempt to compare against unrelated types
        return self.header == other.header
 
    def __hash__(self):
        return hash(self.header)
 
 
# here we can change all_header_defns to a set() object to make sure our header fields are unique, a list would be fine too...
class CSVHeaders:
        all_header_defns = set()
        header_synonyms = []
 
        def __init__(self, csv_headers):
               for csv_header in csv_headers:
                       if csv_header in self.all_header_defns
                               raise CSVHeaderError(f'CSVHeaders object already contains header {csv_header.header}')
                       all_header_defns.add(csv_header)
 
        ...
 
 
# An example of implementing CSVHeaders - define the header_synonyms as a class attribute which auto-generates all of the headers 
# based on the Django model attributes, then when we initialize an instance of PartsListCSVHeaders, it generates all of the header definitions
class PartsListCSVHeaders(CSVHeaders):
        header_synonyms = [
               CSVHeader(header='part_description', header_synonyms=['desc', 'desc.']),
               CSVHeader(header='manufacturer_name', header_synonyms=['mfg_name', 'manufacturer_name', 'part_manufacturer', 'mfg', 'manufacturer', 'manufacturer name', ]),
               ...
               # Stop at tolerance - no need to override with synonyms
        ]
 
    def __init__(self):
        part_headers = CSVHeader.headers_from_model(Part, exclude_fields=['id', 'description', ])
        part_revision_headers = CSVHeader.headers_from_model(PartRevision, exclude_fields=['id', ])
        manufacturer_headers = CSVHeader.headers_from_model(Manufacturer, exclude_fields=['id', 'name', ])
        headers = part_headers + part_revision_headers + manufacturer_headers
        super(CSVHeaders).__init__(csv_headers=headers)

mpkasp commented:
Thanks! All great points. I responded individually, but in summary, I think points #4 and #5 below break a lot of the simplifications I was aiming for. 

I think we can still use a CSVHeader class, but can no longer pull properties from models. There does still look like some low hanging fruit we can simplify, though- specifically the duplication in the class attributes with the instance fields initialized in `__init__()`.

Happy to jump on a call if you'd like, however, I think we're getting pretty close to being on the same page. So far I've just done a read through of the code. I may just need to pull the code in and play with it a bit more at this point, keeping your points in mind. 

1. A CSV column header may have multiple synonymous names.
**Agreed. I think both solutions resolve this.**

2. For a column that is populated from a model object field, at least one of a header’s synonyms should match the corresponding model object field name
**Agreed. I think both solutions resolve this.**

3. No all model object fields are intended to be included in a CVS file.
**Agreed. I think both solutions resolve this. (see the "exclude_fields" parameter in the headers_from_model() classmethod.**

4. The order in which columns are “printed” in a CSV files needs to be controlled. This order generally has nothing to do with the order in which model fields have been declared, or with alphabetization, or anything else that can be inferred. The order of columns is for user consumption (i.e., readability).
**This is a good point that I didn't think of. It may break a lot of the simplifications I was aiming for.**

5. Sometimes the value for a cell in a CSV files is obtained from a model object field, sometimes it is computed. Some of the columns populated from PartBomItem are an example of the latter.
**This is a good point that I didn't think of. It may break a lot of the simplifications I was aiming for.**

6. Not all columns that are imported correspond to model object fields. An example is “synopsis” which is simply ignored when imported.
**Good point.**

robertseliger commented:
I just found another issue with designators and component counts when importing a BOM. The fix will likely only affect forms.py. 

 

I suggest you let me prepare a fix. Then, after I post the fix, you like you can adapt the csv_header,py code as you see fit, or I can make an attempt to integrate the various ideas we have been discussing. Who does what is of course your call to make!

 

In any event, I think it would be helpful if I fixed the designator issue first. I plan to work on it tonight (in fact, I already started.) 

 

Does this make sense? 

 
Rob
 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Monday, February 17, 2020 at 6:30 PM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Author <author@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 

Thanks! All great points. I responded individually, but in summary, I think points #4 and #5 below break a lot of the simplifications I was aiming for.

I think we can still use a CSVHeader class, but can no longer pull properties from models. There does still look like some low hanging fruit we can simplify, though- specifically the duplication in the class attributes with the instance fields initialized in __init__().

Happy to jump on a call if you'd like, however, I think we're getting pretty close to being on the same page. So far I've just done a read through of the code. I may just need to pull the code in and play with it a bit more at this point, keeping your points in mind.
A CSV column header may have multiple synonymous names.
Agreed. I think both solutions resolve this.
For a column that is populated from a model object field, at least one of a header’s synonyms should match the corresponding model object field name
Agreed. I think both solutions resolve this.
No all model object fields are intended to be included in a CVS file.
Agreed. I think both solutions resolve this. (see the "exclude_fields" parameter in the headers_from_model() classmethod.
The order in which columns are “printed” in a CSV files needs to be controlled. This order generally has nothing to do with the order in which model fields have been declared, or with alphabetization, or anything else that can be inferred. The order of columns is for user consumption (i.e., readability).
This is a good point that I didn't think of. It may break a lot of the simplifications I was aiming for.
Sometimes the value for a cell in a CSV files is obtained from a model object field, sometimes it is computed. Some of the columns populated from PartBomItem are an example of the latter.
This is a good point that I didn't think of. It may break a lot of the simplifications I was aiming for.
Not all columns that are imported correspond to model object fields. An example is “synopsis” which is simply ignored when imported.
Good point.
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented:
Sounds good. I'll pull it in and make tweaks as needed. We will probably
need to update tests too.

On Mon, Feb 17, 2020, 6:00 PM robertseliger <notifications@github.com>
wrote:

> I just found another issue with designators and component counts when
> importing a BOM. The fix will likely only affect forms.py.
>
>
>
> I suggest you let me prepare a fix. Then, after I post the fix, you like
> you can adapt the csv_header,py code as you see fit, or I can make an
> attempt to integrate the various ideas we have been discussing. Who does
> what is of course your call to make!
>
>
>
> In any event, I think it would be helpful if I fixed the designator issue
> first. I plan to work on it tonight (in fact, I already started.)
>
>
>
> Does this make sense?
>
>
> Rob
>
>
> From: Mike Kasparian <notifications@github.com>
> Reply-To: mpkasp/django-bom <reply@reply.github.com>
> Date: Monday, February 17, 2020 at 6:30 PM
> To: mpkasp/django-bom <django-bom@noreply.github.com>
> Cc: robertseliger <robert.seliger@whimsmart.com>, Author <
> author@noreply.github.com>
> Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously
> exported CVS files (#34)
>
>
>
> Thanks! All great points. I responded individually, but in summary, I
> think points #4 and #5 below break a lot of the simplifications I was
> aiming for.
>
> I think we can still use a CSVHeader class, but can no longer pull
> properties from models. There does still look like some low hanging fruit
> we can simplify, though- specifically the duplication in the class
> attributes with the instance fields initialized in __init__().
>
> Happy to jump on a call if you'd like, however, I think we're getting
> pretty close to being on the same page. So far I've just done a read
> through of the code. I may just need to pull the code in and play with it a
> bit more at this point, keeping your points in mind.
> A CSV column header may have multiple synonymous names.
> Agreed. I think both solutions resolve this.
> For a column that is populated from a model object field, at least one of
> a header’s synonyms should match the corresponding model object field name
> Agreed. I think both solutions resolve this.
> No all model object fields are intended to be included in a CVS file.
> Agreed. I think both solutions resolve this. (see the "exclude_fields"
> parameter in the headers_from_model() classmethod.
> The order in which columns are “printed” in a CSV files needs to be
> controlled. This order generally has nothing to do with the order in which
> model fields have been declared, or with alphabetization, or anything else
> that can be inferred. The order of columns is for user consumption (i.e.,
> readability).
> This is a good point that I didn't think of. It may break a lot of the
> simplifications I was aiming for.
> Sometimes the value for a cell in a CSV files is obtained from a model
> object field, sometimes it is computed. Some of the columns populated from
> PartBomItem are an example of the latter.
> This is a good point that I didn't think of. It may break a lot of the
> simplifications I was aiming for.
> Not all columns that are imported correspond to model object fields. An
> example is “synopsis” which is simply ignored when imported.
> Good point.
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH54KT5NGS63MYLYESTRDMQKJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMABQCY#issuecomment-587208715>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOOH7AKNVIOFCUHEQSY6TRDMQKJANCNFSM4KWHXW3Q>
> .
>

robertseliger commented:
I will post the fix soon.

For tests, how do we automate CSV file testing without having CSV files?

- Rob

> On Feb 17, 2020, at 7:44 PM, Mike Kasparian <notifications@github.com> wrote:
> 
> ﻿Sounds good. I'll pull it in and make tweaks as needed. We will probably
> need to update tests too.
> 
> On Mon, Feb 17, 2020, 6:00 PM robertseliger <notifications@github.com>
> wrote:
> 
> > I just found another issue with designators and component counts when
> > importing a BOM. The fix will likely only affect forms.py.
> >
> >
> >
> > I suggest you let me prepare a fix. Then, after I post the fix, you like
> > you can adapt the csv_header,py code as you see fit, or I can make an
> > attempt to integrate the various ideas we have been discussing. Who does
> > what is of course your call to make!
> >
> >
> >
> > In any event, I think it would be helpful if I fixed the designator issue
> > first. I plan to work on it tonight (in fact, I already started.)
> >
> >
> >
> > Does this make sense?
> >
> >
> > Rob
> >
> >
> > From: Mike Kasparian <notifications@github.com>
> > Reply-To: mpkasp/django-bom <reply@reply.github.com>
> > Date: Monday, February 17, 2020 at 6:30 PM
> > To: mpkasp/django-bom <django-bom@noreply.github.com>
> > Cc: robertseliger <robert.seliger@whimsmart.com>, Author <
> > author@noreply.github.com>
> > Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously
> > exported CVS files (#34)
> >
> >
> >
> > Thanks! All great points. I responded individually, but in summary, I
> > think points #4 and #5 below break a lot of the simplifications I was
> > aiming for.
> >
> > I think we can still use a CSVHeader class, but can no longer pull
> > properties from models. There does still look like some low hanging fruit
> > we can simplify, though- specifically the duplication in the class
> > attributes with the instance fields initialized in __init__().
> >
> > Happy to jump on a call if you'd like, however, I think we're getting
> > pretty close to being on the same page. So far I've just done a read
> > through of the code. I may just need to pull the code in and play with it a
> > bit more at this point, keeping your points in mind.
> > A CSV column header may have multiple synonymous names.
> > Agreed. I think both solutions resolve this.
> > For a column that is populated from a model object field, at least one of
> > a header’s synonyms should match the corresponding model object field name
> > Agreed. I think both solutions resolve this.
> > No all model object fields are intended to be included in a CVS file.
> > Agreed. I think both solutions resolve this. (see the "exclude_fields"
> > parameter in the headers_from_model() classmethod.
> > The order in which columns are “printed” in a CSV files needs to be
> > controlled. This order generally has nothing to do with the order in which
> > model fields have been declared, or with alphabetization, or anything else
> > that can be inferred. The order of columns is for user consumption (i.e.,
> > readability).
> > This is a good point that I didn't think of. It may break a lot of the
> > simplifications I was aiming for.
> > Sometimes the value for a cell in a CSV files is obtained from a model
> > object field, sometimes it is computed. Some of the columns populated from
> > PartBomItem are an example of the latter.
> > This is a good point that I didn't think of. It may break a lot of the
> > simplifications I was aiming for.
> > Not all columns that are imported correspond to model object fields. An
> > example is “synopsis” which is simply ignored when imported.
> > Good point.
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH54KT5NGS63MYLYESTRDMQKJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMABQCY#issuecomment-587208715>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/ABLOOH7AKNVIOFCUHEQSY6TRDMQKJANCNFSM4KWHXW3Q>
> > .
> >
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented:
Great!

I make fake files, or whenever I find a new bug, I request the file (like I
did once from you!) and use that file and run a test against it.

The files are in the bom/test_files directory

On Mon, Feb 17, 2020 at 6:50 PM robertseliger <notifications@github.com>
wrote:

> I will post the fix soon.
>
> For tests, how do we automate CSV file testing without having CSV files?
>
> - Rob
>
> > On Feb 17, 2020, at 7:44 PM, Mike Kasparian <notifications@github.com>
> wrote:
> >
> > ﻿Sounds good. I'll pull it in and make tweaks as needed. We will probably
> > need to update tests too.
> >
> > On Mon, Feb 17, 2020, 6:00 PM robertseliger <notifications@github.com>
> > wrote:
> >
> > > I just found another issue with designators and component counts when
> > > importing a BOM. The fix will likely only affect forms.py.
> > >
> > >
> > >
> > > I suggest you let me prepare a fix. Then, after I post the fix, you
> like
> > > you can adapt the csv_header,py code as you see fit, or I can make an
> > > attempt to integrate the various ideas we have been discussing. Who
> does
> > > what is of course your call to make!
> > >
> > >
> > >
> > > In any event, I think it would be helpful if I fixed the designator
> issue
> > > first. I plan to work on it tonight (in fact, I already started.)
> > >
> > >
> > >
> > > Does this make sense?
> > >
> > >
> > > Rob
> > >
> > >
> > > From: Mike Kasparian <notifications@github.com>
> > > Reply-To: mpkasp/django-bom <reply@reply.github.com>
> > > Date: Monday, February 17, 2020 at 6:30 PM
> > > To: mpkasp/django-bom <django-bom@noreply.github.com>
> > > Cc: robertseliger <robert.seliger@whimsmart.com>, Author <
> > > author@noreply.github.com>
> > > Subject: Re: [mpkasp/django-bom] Changes to enable importing of
> previously
> > > exported CVS files (#34)
> > >
> > >
> > >
> > > Thanks! All great points. I responded individually, but in summary, I
> > > think points #4 and #5 below break a lot of the simplifications I was
> > > aiming for.
> > >
> > > I think we can still use a CSVHeader class, but can no longer pull
> > > properties from models. There does still look like some low hanging
> fruit
> > > we can simplify, though- specifically the duplication in the class
> > > attributes with the instance fields initialized in __init__().
> > >
> > > Happy to jump on a call if you'd like, however, I think we're getting
> > > pretty close to being on the same page. So far I've just done a read
> > > through of the code. I may just need to pull the code in and play with
> it a
> > > bit more at this point, keeping your points in mind.
> > > A CSV column header may have multiple synonymous names.
> > > Agreed. I think both solutions resolve this.
> > > For a column that is populated from a model object field, at least one
> of
> > > a header’s synonyms should match the corresponding model object field
> name
> > > Agreed. I think both solutions resolve this.
> > > No all model object fields are intended to be included in a CVS file.
> > > Agreed. I think both solutions resolve this. (see the "exclude_fields"
> > > parameter in the headers_from_model() classmethod.
> > > The order in which columns are “printed” in a CSV files needs to be
> > > controlled. This order generally has nothing to do with the order in
> which
> > > model fields have been declared, or with alphabetization, or anything
> else
> > > that can be inferred. The order of columns is for user consumption
> (i.e.,
> > > readability).
> > > This is a good point that I didn't think of. It may break a lot of the
> > > simplifications I was aiming for.
> > > Sometimes the value for a cell in a CSV files is obtained from a model
> > > object field, sometimes it is computed. Some of the columns populated
> from
> > > PartBomItem are an example of the latter.
> > > This is a good point that I didn't think of. It may break a lot of the
> > > simplifications I was aiming for.
> > > Not all columns that are imported correspond to model object fields. An
> > > example is “synopsis” which is simply ignored when imported.
> > > Good point.
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub, or unsubscribe.
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH54KT5NGS63MYLYESTRDMQKJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMABQCY#issuecomment-587208715
> >,
> > > or unsubscribe
> > > <
> https://github.com/notifications/unsubscribe-auth/ABLOOH7AKNVIOFCUHEQSY6TRDMQKJANCNFSM4KWHXW3Q
> >
> > > .
> > >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH55ISF4X2FTXT5S3PDRDMWFJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMAFJAA#issuecomment-587224192>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOOH5COHXXMYXI6IWGENTRDMWFJANCNFSM4KWHXW3Q>
> .
>

robertseliger commented:
I had not noticed the test files directory.

 

I am happy to work on adding tests now that I know that tests can use files. 

 

In the meantime, I just submitted a pull request with updates to forms.py that fix for the designator / count mismatch issue when importing a BOM.

 

I also fixed a comment typo in utils.py (a missing comma), and I changed the CSVHeaders.validate_header_assertions() command name ‘eor’ to ‘me’. The command name ‘me’ stands for mutually exclusive, and this is a more accurate name than ‘eor’, which stood for exclusive-or. I updated the comments in csv_headers.py as well to make this all clear.

 

At this point,  I don’t plan to touch anything other than tests.py, so have at it with fine-tuning csv_headers.py, or if you prefer, tell me what you would me to do.

 
Rob
 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Monday, February 17, 2020 at 8:07 PM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Author <author@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 

Great!

I make fake files, or whenever I find a new bug, I request the file (like I
did once from you!) and use that file and run a test against it.

The files are in the bom/test_files directory

On Mon, Feb 17, 2020 at 6:50 PM robertseliger <notifications@github.com>
wrote:

> I will post the fix soon.
>
> For tests, how do we automate CSV file testing without having CSV files?
>
> - Rob
>
> > On Feb 17, 2020, at 7:44 PM, Mike Kasparian <notifications@github.com>
> wrote:
> >
> > ﻿Sounds good. I'll pull it in and make tweaks as needed. We will probably
> > need to update tests too.
> >
> > On Mon, Feb 17, 2020, 6:00 PM robertseliger <notifications@github.com>
> > wrote:
> >
> > > I just found another issue with designators and component counts when
> > > importing a BOM. The fix will likely only affect forms.py.
> > >
> > >
> > >
> > > I suggest you let me prepare a fix. Then, after I post the fix, you
> like
> > > you can adapt the csv_header,py code as you see fit, or I can make an
> > > attempt to integrate the various ideas we have been discussing. Who
> does
> > > what is of course your call to make!
> > >
> > >
> > >
> > > In any event, I think it would be helpful if I fixed the designator
> issue
> > > first. I plan to work on it tonight (in fact, I already started.)
> > >
> > >
> > >
> > > Does this make sense?
> > >
> > >
> > > Rob
> > >
> > >
> > > From: Mike Kasparian <notifications@github.com>
> > > Reply-To: mpkasp/django-bom <reply@reply.github.com>
> > > Date: Monday, February 17, 2020 at 6:30 PM
> > > To: mpkasp/django-bom <django-bom@noreply.github.com>
> > > Cc: robertseliger <robert.seliger@whimsmart.com>, Author <
> > > author@noreply.github.com>
> > > Subject: Re: [mpkasp/django-bom] Changes to enable importing of
> previously
> > > exported CVS files (#34)
> > >
> > >
> > >
> > > Thanks! All great points. I responded individually, but in summary, I
> > > think points #4 and #5 below break a lot of the simplifications I was
> > > aiming for.
> > >
> > > I think we can still use a CSVHeader class, but can no longer pull
> > > properties from models. There does still look like some low hanging
> fruit
> > > we can simplify, though- specifically the duplication in the class
> > > attributes with the instance fields initialized in __init__().
> > >
> > > Happy to jump on a call if you'd like, however, I think we're getting
> > > pretty close to being on the same page. So far I've just done a read
> > > through of the code. I may just need to pull the code in and play with
> it a
> > > bit more at this point, keeping your points in mind.
> > > A CSV column header may have multiple synonymous names.
> > > Agreed. I think both solutions resolve this.
> > > For a column that is populated from a model object field, at least one
> of
> > > a header’s synonyms should match the corresponding model object field
> name
> > > Agreed. I think both solutions resolve this.
> > > No all model object fields are intended to be included in a CVS file.
> > > Agreed. I think both solutions resolve this. (see the "exclude_fields"
> > > parameter in the headers_from_model() classmethod.
> > > The order in which columns are “printed” in a CSV files needs to be
> > > controlled. This order generally has nothing to do with the order in
> which
> > > model fields have been declared, or with alphabetization, or anything
> else
> > > that can be inferred. The order of columns is for user consumption
> (i.e.,
> > > readability).
> > > This is a good point that I didn't think of. It may break a lot of the
> > > simplifications I was aiming for.
> > > Sometimes the value for a cell in a CSV files is obtained from a model
> > > object field, sometimes it is computed. Some of the columns populated
> from
> > > PartBomItem are an example of the latter.
> > > This is a good point that I didn't think of. It may break a lot of the
> > > simplifications I was aiming for.
> > > Not all columns that are imported correspond to model object fields. An
> > > example is “synopsis” which is simply ignored when imported.
> > > Good point.
> > > —
> > > You are receiving this because you authored the thread.
> > > Reply to this email directly, view it on GitHub, or unsubscribe.
> > >
> > > —
> > > You are receiving this because you commented.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH54KT5NGS63MYLYESTRDMQKJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMABQCY#issuecomment-587208715
> >,
> > > or unsubscribe
> > > <
> https://github.com/notifications/unsubscribe-auth/ABLOOH7AKNVIOFCUHEQSY6TRDMQKJANCNFSM4KWHXW3Q
> >
> > > .
> > >
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH55ISF4X2FTXT5S3PDRDMWFJA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMAFJAA#issuecomment-587224192>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOOH5COHXXMYXI6IWGENTRDMWFJANCNFSM4KWHXW3Q>
> .
>


—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented on bom/views/views.py:
```diff
@@ -655,18 +684,26 @@ def export_part_list(request):
+                if field_name not in csv_headers.get_defaults_list(
+                        ['part_number', 'part_category', 'part_synopsis', 'part_revision', 'part_manufacturer',
+                         'part_manufacturer_part_number', ]):
+                    attr = getattr(part_rev, field_name)
```
@robertseliger looks like this line is failing one of the tests

mpkasp commented:
Looks like there are a few tests failing. Happy to dive into it tonight or tomorrow but it may be quicker if you know the code. Let me know if you're planning on diving into it!

robertseliger commented:
Will take a look in the morning.

- Rob

> On Feb 18, 2020, at 10:10 PM, Mike Kasparian <notifications@github.com> wrote:
> 
> ﻿
> Looks like there are a few tests failing. Happy to dive into it tonight or tomorrow but it may be quicker if you know the code. Let me know if you're planning on diving into it!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

robertseliger commented:
Lots of tweaks in the code and the tests to get all but two tests to pass. In order to get the tests on line 117 and 454 to pass, we need to either change the test code or change the CSV test files. I didn’t know which approach you wanted to take.

 

I will post the changes shortly. See details below

 
Rob
 

 

Tests.py line 117:

 

            The code is properly catching “junk” as an invalid header. The test is needs to be changed, or the header fixed.

 

            part_number,quantity,manufacturer_part_number,dnp,reference,junk

 

tests.py line 454:

 

            The code is properly catching various mistakes in the header, the first of which is that “OC_FARNELL” is not a valid column name.

The test is needs to be changed, or the header fixed.

 

"Part";"Value";"Device";"Package";"Description";"MF";"MPN";"OC_FARNELL";"OC_NEWARK";"PART_CLASS";"PROD_ID";"REVISION";"VALUE";

 

tests.py line 493:

 

The test was looking for the wrong message text. Changed to:

 

self.assertTrue('Part class &#39;code&#39; in row 3 does not have a value. Uploading of this part class skipped.' in str(response.content))

 

tests.py line 545:

            

Deleted this assertion as it no longer makes sense

 

tests.py line 547:

 

            Fixed designator assertions, added subpart count assertions

 

 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Tuesday, February 18, 2020 at 10:10 PM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Mention <mention@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 

Looks like there are a few tests failing. Happy to dive into it tonight or tomorrow but it may be quicker if you know the code. Let me know if you're planning on diving into it!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented:
Feedback below as **MK: in bold**. Great work thanks for fixing the tests.


**Tests.py line 117:**
The code is properly catching “junk” as an invalid header. The test is needs to be changed, or the header fixed.

            part_number,quantity,manufacturer_part_number,dnp,reference,junk

**MK: I think it makes most sense to either change this to a warning, or not an error at all. I'm ok with either and will let you make the call if you agree!**

**tests.py line 454:**
The code is properly catching various mistakes in the header, the first of which is that “OC_FARNELL” is not a valid column name.

The test is needs to be changed, or the header fixed.

"Part";"Value";"Device";"Package";"Description";"MF";"MPN";"OC_FARNELL";"OC_NEWARK";"PART_CLASS";"PROD_ID";"REVISION";"VALUE";

**MK: Same as above - I think it makes most sense to either change this to a warning, or not an error at all. I'm ok with either and will let you make the call if you agree!**

**tests.py line 493:**
The test was looking for the wrong message text. Changed to:

self.assertTrue('Part class &#39;code&#39; in row 3 does not have a value. Uploading of this part class skipped.' in str(response.content))

**MK: Great.**

**tests.py line 545:**
Deleted this assertion as it no longer makes sense
**MK: Great**

tests.py line 547:
            Fixed designator assertions, added subpart count assertions
**MK: Great. I like the addition of subpart count assertions!**

robertseliger commented:
I changed the code so that unrecognized headers now elicit a warning instead of an error are otherwise ignored. I have not posted these changes yet because I encountered a use case I have not thought about ... 

 

Imagine a CSV with a row designated as R1 and a part number 102-00008-01. Imagine that there is another row in the BOM designated as R2 and with the same part number as R1. In the current code, when uploaded, the code will consolidate these two rows in a single subpart with the designator ‘R1, R2’ and a count of 2. Great!

 

But, what do we want to happen if R2 is marked as do-not-load?  I think the answer should be that two subparts are created in the BOM so that the DNL flags can be set independently. The coding for this is tricky because BOMCSVForm.clean() in forms.py is already pretty complicated. I suggest we discuss how to code this capability.

 

I you like I can post the current changes. All tests pass except for one failure in test_part_upload_bom, which is the test for the DNL flag in the scenario I described above.

 
Rob
 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Wednesday, February 19, 2020 at 10:26 PM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Mention <mention@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 

Feedback below as MK: in bold. Great work thanks for fixing the tests.

Tests.py line 117:
The code is properly catching “junk” as an invalid header. The test is needs to be changed, or the header fixed.
        part_number,quantity,manufacturer_part_number,dnp,reference,junk
 
MK: I think it makes most sense to either change this to a warning, or not an error at all. I'm ok with either and will let you make the call if you agree!

tests.py line 454:
The code is properly catching various mistakes in the header, the first of which is that “OC_FARNELL” is not a valid column name.

The test is needs to be changed, or the header fixed.

"Part";"Value";"Device";"Package";"Description";"MF";"MPN";"OC_FARNELL";"OC_NEWARK";"PART_CLASS";"PROD_ID";"REVISION";"VALUE";

MK: Same as above - I think it makes most sense to either change this to a warning, or not an error at all. I'm ok with either and will let you make the call if you agree!

tests.py line 493:
The test was looking for the wrong message text. Changed to:

self.assertTrue('Part class 'code' in row 3 does not have a value. Uploading of this part class skipped.' in str(response.content))

MK: Great.

tests.py line 545:
Deleted this assertion as it no longer makes sense
MK: Great

tests.py line 547:
Fixed designator assertions, added subpart count assertions
MK: Great. I like the addition of subpart count assertions!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented:
Great point. Turns out I actually already handled a scenario just like this
for when a user submits a new subpart to a BOM manually (As opposed to via
a BOM upload). On that note, maybe this logic should be shared between the
two. Either way...

Check out the test `test_add_subpart` - in line 399 there's an assertion:
`self.assertEqual(2, parts_p2)` - this is a check to make sure the part
revision is created twice - once as DNL = False, and once as DNL = True.
The check occurs in the view `add_subpart` and starts on line 875 when we
query the BOM for any subparts with the same part revision, AND DNL status.

Do you want to look into applying similar logic? I'm happy to look into it
as well. Just let me know, but will need your latest changes pushed!

Mike

On Thu, Feb 20, 2020 at 4:11 PM robertseliger <notifications@github.com>
wrote:

> I changed the code so that unrecognized headers now elicit a warning
> instead of an error are otherwise ignored. I have not posted these changes
> yet because I encountered a use case I have not thought about ...
>
>
>
> Imagine a CSV with a row designated as R1 and a part number 102-00008-01.
> Imagine that there is another row in the BOM designated as R2 and with the
> same part number as R1. In the current code, when uploaded, the code will
> consolidate these two rows in a single subpart with the designator ‘R1, R2’
> and a count of 2. Great!
>
>
>
> But, what do we want to happen if R2 is marked as do-not-load?  I think
> the answer should be that two subparts are created in the BOM so that the
> DNL flags can be set independently. The coding for this is tricky because
> BOMCSVForm.clean() in forms.py is already pretty complicated. I suggest we
> discuss how to code this capability.
>
>
>
> I you like I can post the current changes. All tests pass except for one
> failure in test_part_upload_bom, which is the test for the DNL flag in the
> scenario I described above.
>
>
> Rob
>
>
> From: Mike Kasparian <notifications@github.com>
> Reply-To: mpkasp/django-bom <reply@reply.github.com>
> Date: Wednesday, February 19, 2020 at 10:26 PM
> To: mpkasp/django-bom <django-bom@noreply.github.com>
> Cc: robertseliger <robert.seliger@whimsmart.com>, Mention <
> mention@noreply.github.com>
> Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously
> exported CVS files (#34)
>
>
>
> Feedback below as MK: in bold. Great work thanks for fixing the tests.
>
> Tests.py line 117:
> The code is properly catching “junk” as an invalid header. The test is
> needs to be changed, or the header fixed.
>         part_number,quantity,manufacturer_part_number,dnp,reference,junk
>
> MK: I think it makes most sense to either change this to a warning, or not
> an error at all. I'm ok with either and will let you make the call if you
> agree!
>
> tests.py line 454:
> The code is properly catching various mistakes in the header, the first of
> which is that “OC_FARNELL” is not a valid column name.
>
> The test is needs to be changed, or the header fixed.
>
>
> "Part";"Value";"Device";"Package";"Description";"MF";"MPN";"OC_FARNELL";"OC_NEWARK";"PART_CLASS";"PROD_ID";"REVISION";"VALUE";
>
> MK: Same as above - I think it makes most sense to either change this to a
> warning, or not an error at all. I'm ok with either and will let you make
> the call if you agree!
>
> tests.py line 493:
> The test was looking for the wrong message text. Changed to:
>
> self.assertTrue('Part class 'code' in row 3 does not have a value.
> Uploading of this part class skipped.' in str(response.content))
>
> MK: Great.
>
> tests.py line 545:
> Deleted this assertion as it no longer makes sense
> MK: Great
>
> tests.py line 547:
> Fixed designator assertions, added subpart count assertions
> MK: Great. I like the addition of subpart count assertions!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH2U426HIAWI5OPPFXDRD352PA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMQSJIA#issuecomment-589374624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOOH5HKW3V3AZPEUUFDDDRD352PANCNFSM4KWHXW3Q>
> .
>

robertseliger commented:
I will look at the code that deals with the test on line 399. At the moment, the code I am working on handles the tests around lines 130. 

 
Rob
 

From: Mike Kasparian <notifications@github.com>
Reply-To: mpkasp/django-bom <reply@reply.github.com>
Date: Saturday, February 22, 2020 at 9:28 AM
To: mpkasp/django-bom <django-bom@noreply.github.com>
Cc: robertseliger <robert.seliger@whimsmart.com>, Mention <mention@noreply.github.com>
Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously exported CVS files (#34)

 

Great point. Turns out I actually already handled a scenario just like this
for when a user submits a new subpart to a BOM manually (As opposed to via
a BOM upload). On that note, maybe this logic should be shared between the
two. Either way...

Check out the test `test_add_subpart` - in line 399 there's an assertion:
`self.assertEqual(2, parts_p2)` - this is a check to make sure the part
revision is created twice - once as DNL = False, and once as DNL = True.
The check occurs in the view `add_subpart` and starts on line 875 when we
query the BOM for any subparts with the same part revision, AND DNL status.

Do you want to look into applying similar logic? I'm happy to look into it
as well. Just let me know, but will need your latest changes pushed!

Mike

On Thu, Feb 20, 2020 at 4:11 PM robertseliger <notifications@github.com>
wrote:

> I changed the code so that unrecognized headers now elicit a warning
> instead of an error are otherwise ignored. I have not posted these changes
> yet because I encountered a use case I have not thought about ...
>
>
>
> Imagine a CSV with a row designated as R1 and a part number 102-00008-01.
> Imagine that there is another row in the BOM designated as R2 and with the
> same part number as R1. In the current code, when uploaded, the code will
> consolidate these two rows in a single subpart with the designator ‘R1, R2’
> and a count of 2. Great!
>
>
>
> But, what do we want to happen if R2 is marked as do-not-load? I think
> the answer should be that two subparts are created in the BOM so that the
> DNL flags can be set independently. The coding for this is tricky because
> BOMCSVForm.clean() in forms.py is already pretty complicated. I suggest we
> discuss how to code this capability.
>
>
>
> I you like I can post the current changes. All tests pass except for one
> failure in test_part_upload_bom, which is the test for the DNL flag in the
> scenario I described above.
>
>
> Rob
>
>
> From: Mike Kasparian <notifications@github.com>
> Reply-To: mpkasp/django-bom <reply@reply.github.com>
> Date: Wednesday, February 19, 2020 at 10:26 PM
> To: mpkasp/django-bom <django-bom@noreply.github.com>
> Cc: robertseliger <robert.seliger@whimsmart.com>, Mention <
> mention@noreply.github.com>
> Subject: Re: [mpkasp/django-bom] Changes to enable importing of previously
> exported CVS files (#34)
>
>
>
> Feedback below as MK: in bold. Great work thanks for fixing the tests.
>
> Tests.py line 117:
> The code is properly catching “junk” as an invalid header. The test is
> needs to be changed, or the header fixed.
> part_number,quantity,manufacturer_part_number,dnp,reference,junk
>
> MK: I think it makes most sense to either change this to a warning, or not
> an error at all. I'm ok with either and will let you make the call if you
> agree!
>
> tests.py line 454:
> The code is properly catching various mistakes in the header, the first of
> which is that “OC_FARNELL” is not a valid column name.
>
> The test is needs to be changed, or the header fixed.
>
>
> "Part";"Value";"Device";"Package";"Description";"MF";"MPN";"OC_FARNELL";"OC_NEWARK";"PART_CLASS";"PROD_ID";"REVISION";"VALUE";
>
> MK: Same as above - I think it makes most sense to either change this to a
> warning, or not an error at all. I'm ok with either and will let you make
> the call if you agree!
>
> tests.py line 493:
> The test was looking for the wrong message text. Changed to:
>
> self.assertTrue('Part class 'code' in row 3 does not have a value.
> Uploading of this part class skipped.' in str(response.content))
>
> MK: Great.
>
> tests.py line 545:
> Deleted this assertion as it no longer makes sense
> MK: Great
>
> tests.py line 547:
> Fixed designator assertions, added subpart count assertions
> MK: Great. I like the addition of subpart count assertions!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mpkasp/django-bom/pull/34?email_source=notifications&email_token=ABLOOH2U426HIAWI5OPPFXDRD352PA5CNFSM4KWHXW32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMQSJIA#issuecomment-589374624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABLOOH5HKW3V3AZPEUUFDDDRD352PANCNFSM4KWHXW3Q>
> .
>


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.

mpkasp commented:
High level looks good & tests are passing. Not able to do a deep dive but I want to get this in so I'll work on merging it.
